File: configuration/src/test/java/com/synopsys/integration/configuration/property/PropertyTestHelpUtil.java
package com.synopsys.integration.configuration.property;

import java.util.List;

import org.apache.commons.collections4.CollectionUtils;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Assertions;

import com.synopsys.integration.configuration.property.base.NullableProperty;
import com.synopsys.integration.configuration.property.base.TypedProperty;
import com.synopsys.integration.configuration.property.base.ValuedListProperty;
import com.synopsys.integration.configuration.property.base.ValuedProperty;

/**
 * Ensures the Property classes are providing necessary property descriptions for help doc.
 */
public class PropertyTestHelpUtil {

    private PropertyTestHelpUtil() {
        throw new IllegalStateException("Utility class");
    }

    //#region Recommended Usage

    public static <T, V> void assertAllHelpValid(NullableProperty<T, V> property) {
        assertAllHelpValid(property, null);
    }

    public static <T, V> void assertAllHelpValid(NullableProperty<T, V> property, @Nullable List<String> expectedExampleValues) {
        assertValidTypeDescription(property);
        assertHasExampleValues(property, expectedExampleValues);
    }

    public static <T, V> void assertAllHelpValid(ValuedProperty<T, V> property) {
        assertAllHelpValid(property, null);
    }

    public static <T, V> void assertAllHelpValid(ValuedProperty<T, V> property, @Nullable List<String> expectedExampleValues) {
        assertValidTypeDescription(property);
        assertHasDefaultDescription(property);
        assertHasExampleValues(property, expectedExampleValues);
    }

    public static <T, V> void assertAllListHelpValid(ValuedListProperty<T, V> property) {
        assertAllListHelpValid(property, null);
    }

    public static <T, V> void assertAllListHelpValid(ValuedListProperty<T, V> property, @Nullable List<String> expectedExampleValues) {
        assertValidTypeDescription(property);
        assertHasDefaultDescription(property);
        assertHasExampleValues(property, expectedExampleValues);
    }

    //#endregion Recommended Usage

    //#region Advanced Usage

    private static final String NULLABLE_TYPE_DESCRIPTION_PREFIX = "Optional ";
    private static final String VALUED_TYPE_DESCRIPTION_POSTFIX = " List";

    public static <T, V> void assertValidTypeDescription(NullableProperty<T, V> property) {
        assertHasTypeDescription(property);
        Assertions.assertTrue(
            property.describeType().startsWith(NULLABLE_TYPE_DESCRIPTION_PREFIX),
            String.format(
                "%s is a %s so its type description should start with '%s'.",
                property.getClass().getSimpleName(),
                NullableProperty.class.getSimpleName(),
                NULLABLE_TYPE_DESCRIPTION_PREFIX
            )
        );
    }

    public static <T, V> void assertValidTypeDescription(ValuedProperty<T, V> property) {
        assertHasTypeDescription(property);
        Assertions.assertFalse(
            property.describeType().startsWith(NULLABLE_TYPE_DESCRIPTION_PREFIX),
            String.format(
                "%s is a %s so its type description should not start with '%s'.",
                property.getClass().getSimpleName(),
                ValuedProperty.class.getSimpleName(),
                NULLABLE_TYPE_DESCRIPTION_PREFIX
            )
        );
    }

    public static <T, V> void assertValidTypeDescription(ValuedListProperty<T, V> property) {
        assertHasTypeDescription(property);
        Assertions.assertTrue(
            property.describeType().endsWith(VALUED_TYPE_DESCRIPTION_POSTFIX),
            String.format(
                "%s is a %s so its type description should end with '%s'.",
                property.getClass().getSimpleName(),
                ValuedListProperty.class.getSimpleName(),
                VALUED_TYPE_DESCRIPTION_POSTFIX
            )
        );
    }

    private static <T, V> void assertHasTypeDescription(TypedProperty<T, V> property) {
        Assertions.assertNotNull(property.describeType(), String.format("%s is a typed property so it should be able to describe its type.", property.getClass().getSimpleName()));
    }

    public static <T, V> void assertHasExampleValues(TypedProperty<T, V> property, @Nullable List<String> expectedExampleValues) {
        if (expectedExampleValues != null) {
            Assertions.assertNotNull(
                property.listExampleValues(),
                String.format("A %s property has a limited number of values that should be described.", property.describeType())
            );
            //Ideally would use a CollectionUtils.areEqual but this gets us close enough.
            Assertions.assertEquals(
                expectedExampleValues.size(),
                property.listExampleValues().size(),
                String.format("The %s provided unexpected example values.", property.getClass().getSimpleName())
            );
            Assertions.assertTrue(
                CollectionUtils.containsAll(expectedExampleValues, property.listExampleValues()),
                String.format("The %s provided unexpected example values.", property.getClass().getSimpleName())
            );
            Assertions.assertTrue(
                CollectionUtils.containsAll(property.listExampleValues(), expectedExampleValues),
                String.format("The %s provided unexpected example values.", property.getClass().getSimpleName())
            );
        }
    }

    private static <T, V> void assertHasExampleValues(ValuedListProperty<T, V> property, @Nullable List<String> expectedExampleValues) {
        if (expectedExampleValues != null) {
            Assertions.assertNotNull(
                property.listExampleValues(),
                String.format("A %s property has a limited number of values that should be described.", property.describeType())
            );
            Assertions.assertTrue(
                CollectionUtils.containsAll(expectedExampleValues, property.listExampleValues()),
                String.format("The %s provided unexpected example values.", property.getClass().getSimpleName())
            );
        }
    }

    public static <T, V> void assertHasDefaultDescription(ValuedProperty<T, V> property) {
        Assertions.assertNotNull(
            property.describeDefault(),
            String.format("%s is a non-null typed property and should describe its default value.", property.getClass().getSimpleName())
        );
    }

    //#endregion Advanced Usage
}

File: configuration/src/test/java/com/synopsys/integration/configuration/property/PropertyTestHelpUtil.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.property

import com.synopsys.integration.configuration.property.base.NullableProperty
import com.synopsys.integration.configuration.property.base.TypedProperty
import com.synopsys.integration.configuration.property.base.ValuedListProperty
import com.synopsys.integration.configuration.property.base.ValuedProperty
import org.junit.jupiter.api.Assertions

/**
 * Ensure's the Property classes are providing necessary property descriptions for help doc.
 */
class PropertyTestHelpUtil {
    companion object {
        //#region Recommended Usage

        fun assertAllHelpValid(property: NullableProperty<*>, expectedExampleValues: List<String>? = null) {
            assertValidTypeDescription(property)
            assertHasExampleValues(property, expectedExampleValues)
        }

        fun assertAllHelpValid(property: ValuedProperty<*>, expectedExampleValues: List<String>? = null) {
            assertValidTypeDescription(property)
            assertHasExampleValues(property, expectedExampleValues)
        }

        fun assertAllHelpValid(property: ValuedListProperty<*>, expectedExampleValues: List<String>? = null) {
            assertValidTypeDescription(property)
            assertHasDefaultDescription(property)
            assertHasExampleValues(property, expectedExampleValues)
        }

        //#endregion Recommended Usage

        //#region Advanced Usage

        private const val nullableTypeDescriptionPrefix = "Optional "
        private const val valuedTypeDescriptionPostfix = " List"

        fun assertValidTypeDescription(property: NullableProperty<*>) {
            assertHasTypeDescription(property)
            Assertions.assertTrue(property.describeType()!!.startsWith(nullableTypeDescriptionPrefix), "${property.javaClass.simpleName} is a ${NullableProperty::class.java.simpleName} so its type description should start with '$nullableTypeDescriptionPrefix'.")
        }

        fun assertValidTypeDescription(property: ValuedProperty<*>) {
            assertHasTypeDescription(property)
            Assertions.assertFalse(property.describeType()!!.startsWith(nullableTypeDescriptionPrefix), "${property.javaClass.simpleName} is a ${ValuedProperty::class.java.simpleName} so its type description should not start with '$nullableTypeDescriptionPrefix'.")
            Assertions.assertFalse(property.describeType()!!.endsWith(valuedTypeDescriptionPostfix), "${property.javaClass.simpleName} is a ${ValuedProperty::class.java.simpleName} so its type description should not start with '$valuedTypeDescriptionPostfix'.")
        }

        fun assertValidTypeDescription(property: ValuedListProperty<*>) {
            assertHasTypeDescription(property)
            Assertions.assertTrue(property.describeType()!!.endsWith(valuedTypeDescriptionPostfix), "${property.javaClass.simpleName} is a ${ValuedListProperty::class.java.simpleName} so its type description should end with '$valuedTypeDescriptionPostfix'.")
            Assertions.assertNotNull(property.describeType(), "${property.javaClass.simpleName} is a typed property such and should describe its type.")
        }

        private fun assertHasTypeDescription(property: TypedProperty<*>) {
            Assertions.assertNotNull(property.describeType(), "${property.javaClass.simpleName} is a typed property such and should describe its type.")
        }

        fun assertHasExampleValues(property: TypedProperty<*>, expectedExampleValues: List<String>?) {
            Assertions.assertNotNull(property.listExampleValues(), "A ${property.describeType()} property has a limited number of values that should be described.")

            if (expectedExampleValues == null) {
                Assertions.assertNotEquals(emptyList<String>(), property.listExampleValues(), "${property.javaClass.simpleName} is typed and therefore should provide example values.")
            } else {
                Assertions.assertEquals(expectedExampleValues.sorted(), property.listExampleValues()?.sorted(), "The ${property.javaClass.simpleName} provided unexpected example values.")
            }
        }

        fun assertHasDefaultDescription(property: ValuedProperty<*>) {
            Assertions.assertNotNull(property.describeDefault(), "${property.javaClass.simpleName} is a non-null typed property and should describe its default value.")
        }

        //#endregion Advanced Usage
    }
}